### PR TITLE
Fix memory leaks in kv_storage and device creation

### PR DESCRIPTION
**Pull request by sorru94**

### CODE REVIEW

**Review by lucaato — CHANGES_REQUESTED** [review]
Everything looks good apart from some doubts about the tests.

File: lib/astarte_device_sdk/device_connection.c [hunk 1]
```diff
@@ -216,12 +216,20 @@ static void setup_subscriptions(astarte_device_handle_t device)
...
-                device->device_id, interface->name);
-            if (ret < 0) {
-                ASTARTE_LOG_ERR("Error encoding MQTT topic");
-                ASTARTE_LOG_ERR("Might be out of memory %s: %d", __FILE__, __LINE__);
+            size_t topic_len = strlen(CONFIG_ASTARTE_DEVICE_SDK_REALM_NAME "///#")
+                + ASTARTE_DEVICE_ID_LEN + strlen(interface->name);
+            char *topic = calloc(topic_len + 1, sizeof(char));
+            if (!topic) {
+                ASTARTE_LOG_ERR("Out of memory %s: %d", __FILE__, __LINE__);
+                continue;
```
[the remark this answers is not in the document]
Comment (sorru94) [reply lib/astarte_device_sdk/device_connection.c:224]: This is a tricky issue. Let's discuss it IRL.

File: lib/astarte_device_sdk/device.c [hunk 2]
```diff
@@ -174,15 +173,21 @@ astarte_result_t astarte_device_new(astarte_device_config_t *cfg, astarte_device
     return ares;
 
 failure:
+    if (handle && handle->introspection.list) {
+        introspection_free(handle->introspection);
```
Comment (lucaato) [inline lib/astarte_device_sdk/device.c:177]: One last thing, I'm not sure if checking the list field from external code is a good practice. We should probably move the if inside the introspection free function and check only the handle here.

Comment (harlem88) [reply lib/astarte_device_sdk/device.c:177]: mmm, Is it right to check `handle->introspection.list` and not  `handle->introspection`?

Comment (sorru94) [reply lib/astarte_device_sdk/device.c:177]: Sure, let's check it inside.

**Review by lucaato — APPROVED** [review]

**Review by harlem88 — APPROVED** [review]

**Review by harlem88 — APPROVED** [review]

**Pull request merged**